### PR TITLE
x64: Add non-SSSE3 lowerings of `pshufb`

### PR DESCRIPTION
**Pull request by alexcrichton**
Or, more accurately, add lowerings which don't use `pshufb`'s functionality at all where possible or otherwise fall back to a new libcall. This particular instruction seemed uniquely difficult to implement in the backend so I decided to "cop out" and use libcall instead. The libcall will be used for `popcnt`, `shuffle`, and `swizzle` instructions when SSSE3 isn't available.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This ended up being a bit more complicated than I originally thought. Turns out that Cranelift previously didn't correctly implement the Windows fastcall calling convention with respect to vector arguments since they're all supposed to be indirect rather than immediate in registers (unlike System-V). That motivated me to remove the usage of `lower_pshufb` and the usage of a libcall, ending up in an SSE2 lowering for `i8x16.popcnt` that's not the end of the world. I couldn't find a way to avoid the libcall for `shuffle` or `swizzle`, though, so I doubled down on fixing it. I've attempted to add the ABI bits here to get that working, but more eyes are of course always appreciated.

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 1]
```diff
@@ -154,6 +154,42 @@ impl ABIMachineSpec for X64ABIMachineSpec {
...
+                            ty: ir::types::I64,
+                            extension: ir::ArgumentExtension::None,
+                        }
+                    }
+
+                    // FIXME: this is implementable but is left for a future
+                    // commit. The pointer needs to become a stack argument, but
+                    // all the stack reservation for the actual values probably
+                    // needs to come before all the argument stack area, so
+                    // that refactoring is left for later.
```
Comment (fitzgen) [inline cranelift/codegen/src/isa/x64/abi.rs]: File a follow up issue to track this?

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 2]
```diff
@@ -89,7 +89,7 @@ impl ABIMachineSpec for X64ABIMachineSpec {
         mut args: ArgsAccumulator<'_>,
     ) -> CodegenResult<(u32, Option<usize>)>
     where
-        I: IntoIterator<Item = &'a ir::AbiParam>,
+        I: IntoIterator<Item = &'a ir::AbiParam> + Clone,
```
Comment (fitzgen) [inline cranelift/codegen/src/isa/x64/abi.rs:92]: Is this still necessary? I'm not seeing any `.clone()`s here.

**Review by fitzgen — APPROVED** [review]

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 2 above)
Comment (alexcrichton) [reply cranelift/codegen/src/isa/x64/abi.rs]: Oops, indeed!

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 1 above)
Comment (alexcrichton) [reply cranelift/codegen/src/isa/x64/abi.rs]: Ah this ended up actually being necessary to fill out, so no longer relevant (sorry I probably pushed to the PR while you were reviewing)

**Pull request merged**